pkgs/instrumentations/web/page-view/src/instrumentation.ts

```diff
@@ -0,0 +1,187 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  InstrumentationBase,
+  isWrapped,
+} from '@opentelemetry/sandbox-instrumentation';
+import { Event } from '@opentelemetry/sandbox-api-events';
+import { EventEmitter } from '@opentelemetry/sandbox-sdk-events';
+
+import { VERSION } from './version';
+import {
+  ApplyCustomEventDataFunction,
+  PageViewInstrumentationConfig,
+} from './types';
+import { PageTypes } from './enums/PageTypes';
+/**
+ * This class represents a document load plugin
+ */
+
+export class PageViewEventInstrumentation extends InstrumentationBase<unknown> {
+  static readonly instrumentationName =
+    '@opentelemetry/instrumentation-page-view';
+  readonly component: string = 'page-view-event';
+  readonly version: string = '1';
+  moduleName = this.component;
+  emitter: EventEmitter | null = null;
+  oldUrl = location.href;
+  applyCustomEventData: ApplyCustomEventDataFunction | undefined = undefined;
+
+  /**
+   *
+   * @param config
+   */
+  constructor(config: PageViewInstrumentationConfig) {
+    super(PageViewEventInstrumentation.instrumentationName, VERSION, config);
+    this.emitter = new EventEmitter(
+      PageViewEventInstrumentation.name,
+      VERSION,
+      'browser'
+    );
+    this.applyCustomEventData = config.applyCustomEventData;
+    this._patchHistoryApi();
+  }
+
+  init() {}
+
+  /**
+   * callback to be executed when page is viewed
+   */
+  private _onPageView() {
+    const pageViewEvent: Event = {
+      name: 'page_view',
+      data: {
+        url: document.documentURI as string,
+        referrer: document.referrer,
+        title: document.title,
+        type: PageTypes.BASE_PAGE,
+      },
+    };
+    this._applyCustomEventData(pageViewEvent, this.applyCustomEventData);
+    this.emitter?.emit(pageViewEvent);
+  }
+
+  /**
+   * callback to be executed when page is viewed
+   */
+  private _onVirtualPageView(changeState: string | null | undefined) {
+    const title = document.title;
+    const referrer = this.oldUrl;
+    // Dont emit an event if the route didn't change
+    if (referrer === location.href) {
+      return;
+    }
+    const vPageViewEvent: Event = {
+      name: 'page_view',
+      data: {
+        'http.url': window.location.href,
+        title,
+        changeState: changeState || '',
+        referrer,
+        type: PageTypes.VIRTUTAL_PAGE,
+      },
+    };
+    this._applyCustomEventData(vPageViewEvent, this.applyCustomEventData);
+    this.emitter?.emit(vPageViewEvent);
+  }
+
+  public _setEmitter(emitter: EventEmitter) {
+    this.emitter = emitter;
+  }
+
+  /**
+   * executes callback {_onDocumenContentLoaded } when the page is viewed
+   */
+  private _waitForPageLoad() {
+    document.addEventListener('DOMContentLoaded', this._onPageView.bind(this));
```

Comment:
I think this should be the document load event instead

Reply:
I thought this event should be fired as soon as a page view happened, which is better taken as soon as the dom contnent(html) is loaded instead of waiting until all resources to are loaded , and the dom is rendered